### PR TITLE
Improve generator.py network support, and some cleanup

### PR DESCRIPTION
**Pull request by kormat**
- Remove ":"s from topology path and file names, for better readability.
- Remove unneeded directory creation code.
- Add lib.util.copy_file() to create dirs as needed when copying.
- Cleaned up the ADConfigurations.json format a bit:
  - ADs are no longer top-level entries.
  - Moved default values into a top-level 'defaults' category.
- Moved init code from generate_all into init.
- Stop passing the ad config dict to every single method.
- Replaced path_dict method with small utility methods.
- Network can be passed in on the command-line, and over-rides the
  ADConfigurations setting.
- Replaced network addressing code and set_er_ip_addresses with a simple
  set of factory classes.
  - Per-ad subnets in the config file weren't used, so support for this
    is dropped, for now at least.
- Use temp directory for TRC temporary file.
  <a href='#crh-start'></a><a href='#crh-data-%7B%22approved%22%3A%20%7B%22https%3A//github.com/shitz%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3840858%3Fv%3D3%22%7D%7D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/415%23issuecomment-147425705%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%3Ashipit%3A%20%22%2C%20%22created_at%22%3A%20%222015-10-12T15%3A02%3A02Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3840858%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/shitz%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%20131b4b30d1657278de6b3e8c76fde370a1666bcf%20topology/generator.py%20991%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/415%23discussion_r41764543%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22This%20should%20raise%20an%20exception%20when%20the%20generator%20cannot%20allocate%20more%20addresses%20within%20the%20subnet.%22%2C%20%22created_at%22%3A%20%222015-10-12T14%3A54%3A45Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3840858%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/shitz%22%7D%7D%2C%20%7B%22body%22%3A%20%22I%27ve%20added%20error%20logging%20for%20these%20cases%3A%5Cr%5Cn%5Cr%5CnCRITICAL%3Aroot%3ANetwork%2010.0.0.0/31%20is%20too%20small%20to%20accomadate%20/27%20subnets%5Cr%5CnCRITICAL%3Aroot%3AUnable%20to%20allocate%20any%20more%20subnets%20from%2010.0.0.0/26%5Cr%5CnCRITICAL%3Aroot%3AUnable%20to%20allocate%20any%20more%20addresses%20from%2010.0.0.0/31%22%2C%20%22created_at%22%3A%20%222015-10-12T15%3A00%3A56Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%20%5Cr%5Cn%22%2C%20%22created_at%22%3A%20%222015-10-12T15%3A01%3A56Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3840858%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/shitz%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20topology/generator.py%3AL782-808%22%7D%2C%20%22Pull%20259b61c9afef32547c3828c2ec05d1c6436c3ac2%20topology/generator.py%20405%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/415%23discussion_r41762985%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22This%20condition%20seems%20a%20bit%20arbitrary%20to%20me.%20I%20don%27t%20think%20we%20should%20_disallow_%20intermediate%20ASes%20to%20have%20path%20servers.%20I%27d%20only%20do%20a%20%60if%20%5C%22path_servers%5C%22%20in%20ad_conf%3A%60%22%2C%20%22created_at%22%3A%20%222015-10-12T14%3A40%3A57Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3840858%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/shitz%22%7D%7D%2C%20%7B%22body%22%3A%20%22I%20would%20prefer%20if%20that%20was%20changed%20in%20a%20different%20PR%20%28it%20also%20probably%20affects%202%20other%20places%20in%20generator.py%29.%22%2C%20%22created_at%22%3A%20%222015-10-12T14%3A44%3A44Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22Ok.%20I%20see%20that%20the%20generator%20needs%20some%20major%20overhaul%20anyway.%20%3A%2B1%3A%20%22%2C%20%22created_at%22%3A%20%222015-10-12T14%3A53%3A21Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3840858%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/shitz%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20topology/generator.py%3AL320-408%22%7D%2C%20%22Pull%20259b61c9afef32547c3828c2ec05d1c6436c3ac2%20topology/generator.py%20132%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/415%23discussion_r41762268%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Can%20we%20name%20this%20%60_write_path_policy_files%28%29%60%20please%3F%20Similar%20for%20%60topo%60%20below.%22%2C%20%22created_at%22%3A%20%222015-10-12T14%3A33%3A53Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3840858%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/shitz%22%7D%7D%2C%20%7B%22body%22%3A%20%22Yeah%20sure.%22%2C%20%22created_at%22%3A%20%222015-10-12T14%3A37%3A05Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%20%22%2C%20%22created_at%22%3A%20%222015-10-12T14%3A53%3A40Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3840858%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/shitz%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20topology/generator.py%3AL116-182%22%7D%2C%20%22Pull%20259b61c9afef32547c3828c2ec05d1c6436c3ac2%20topology/ADConfigurations.json%20206%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/415%23discussion_r41761835%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Is%20there%20a%20reason%20why%20you%20left%20out%20port%20%604005%60%3F%20It%20probably%20doesn%27t%20matter%20at%20all%2C%20but%20it%20looks%20like%20an%20unintentional%20thing%20to%20me.%22%2C%20%22created_at%22%3A%20%222015-10-12T14%3A29%3A41Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3840858%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/shitz%22%7D%7D%2C%20%7B%22body%22%3A%20%22Haha%2C%20nope.%20I%27ll%20change%20it.%5Cr%5Cn%22%2C%20%22created_at%22%3A%20%222015-10-12T14%3A36%3A58Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%20%22%2C%20%22created_at%22%3A%20%222015-10-12T14%3A53%3A47Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3840858%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/shitz%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20topology/ADConfigurations.json%3AL1-177%22%7D%2C%20%22Pull%20259b61c9afef32547c3828c2ec05d1c6436c3ac2%20topology/generator.py%20327%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/415%23discussion_r41762716%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Please%20rename%20to%20%60_write_sim_run_preamble%28%29%60%20or%20something%20like%20that.%22%2C%20%22created_at%22%3A%20%222015-10-12T14%3A38%3A10Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3840858%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/shitz%22%7D%7D%2C%20%7B%22body%22%3A%20%22Done%2C%20gladly.%22%2C%20%22created_at%22%3A%20%222015-10-12T14%3A43%3A47Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%20%22%2C%20%22created_at%22%3A%20%222015-10-12T14%3A53%3A32Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3840858%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/shitz%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20topology/generator.py%3AL271-278%22%7D%7D%2C%20%22processed%22%3A%20%5B%22https%3A//github.com/netsec-ethz/scion/pull/415%23issuecomment-147425705%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/415%23discussion_r41761835%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/415%23discussion_r41762268%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/415%23discussion_r41762597%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/415%23discussion_r41762605%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/415%23discussion_r41762716%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/415%23discussion_r41762985%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/415%23discussion_r41763280%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/415%23discussion_r41763362%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/415%23discussion_r41764387%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/415%23discussion_r41764408%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/415%23discussion_r41764426%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/415%23discussion_r41764441%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/415%23discussion_r41764543%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/415%23discussion_r41765316%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/415%23discussion_r41765438%22%5D%7D'></a>
  <a href='https://www.codereviewhub.com/'></a>

<a href='https://github.com/shitz'></a>
- [x] <a href='#crh-comment-Pull 259b61c9afef32547c3828c2ec05d1c6436c3ac2 topology/ADConfigurations.json 206'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/415#discussion_r41761835'>File: topology/ADConfigurations.json:L1-177</a></b>
- <a href='https://github.com/shitz'></a> Is there a reason why you left out port `4005`? It probably doesn't matter at all, but it looks like an unintentional thing to me.
- <a href='https://github.com/kormat'></a> Haha, nope. I'll change it.
- [x] <a href='#crh-comment-Pull 259b61c9afef32547c3828c2ec05d1c6436c3ac2 topology/generator.py 132'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/415#discussion_r41762268'>File: topology/generator.py:L116-182</a></b>
- <a href='https://github.com/shitz'></a> Can we name this `_write_path_policy_files()` please? Similar for `topo` below.
- <a href='https://github.com/kormat'></a> Yeah sure.
- [x] <a href='#crh-comment-Pull 259b61c9afef32547c3828c2ec05d1c6436c3ac2 topology/generator.py 327'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/415#discussion_r41762716'>File: topology/generator.py:L271-278</a></b>
- <a href='https://github.com/shitz'></a> Please rename to `_write_sim_run_preamble()` or something like that.
- <a href='https://github.com/kormat'></a> Done, gladly.
- [x] <a href='#crh-comment-Pull 259b61c9afef32547c3828c2ec05d1c6436c3ac2 topology/generator.py 405'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/415#discussion_r41762985'>File: topology/generator.py:L320-408</a></b>
- <a href='https://github.com/shitz'></a> This condition seems a bit arbitrary to me. I don't think we should _disallow_ intermediate ASes to have path servers. I'd only do a `if "path_servers" in ad_conf:`
- <a href='https://github.com/kormat'></a> I would prefer if that was changed in a different PR (it also probably affects 2 other places in generator.py).
- <a href='https://github.com/shitz'></a> Ok. I see that the generator needs some major overhaul anyway. :+1:
- [x] <a href='#crh-comment-Pull 131b4b30d1657278de6b3e8c76fde370a1666bcf topology/generator.py 991'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/415#discussion_r41764543'>File: topology/generator.py:L782-808</a></b>
- <a href='https://github.com/shitz'></a> This should raise an exception when the generator cannot allocate more addresses within the subnet.
- <a href='https://github.com/kormat'></a> I've added error logging for these cases:
  CRITICAL:root:Network 10.0.0.0/31 is too small to accomadate /27 subnets
  CRITICAL:root:Unable to allocate any more subnets from 10.0.0.0/26
  CRITICAL:root:Unable to allocate any more addresses from 10.0.0.0/31

<a href='https://www.codereviewhub.com/netsec-ethz/scion/pull/415?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/netsec-ethz/scion/pull/415?approve=1'></a>&nbsp;<a href='https://www.codereviewhub.com/netsec-ethz/scion/pull/415?approve=0'></a>&nbsp;<a href='https://github.com/netsec-ethz/scion/pull/415'></a>
<a href='#crh-end'></a>

### CODE REVIEW

File: topology/ADConfigurations.json [hunk 1]
```diff
@@ -1,172 +1,176 @@
...
+                    "manage": true,
+                    "client_port": 4000,
+                    "leader_port": 4001,
+                    "election_port": 4002
+                },
+                "2": {
+                    "manage": true,
+                    "client_port": 4003,
+                    "leader_port": 4004,
+                    "election_port": 4006
```
Comment (shitz) [inline topology/ADConfigurations.json]: Is there a reason why you left out port `4005`? It probably doesn't matter at all, but it looks like an unintentional thing to me.

File: topology/generator.py [hunk 2]
```diff
@@ -110,80 +116,66 @@ def __init__(self, out_dir=TOPOLOGY_PATH, subnet=DEFAULT_SUBNET,
...
-                                          file_no_ext + '.json')
-        path_pol_file_abs = os.path.join(self.out_dir, path_pol_path_tail)
-        path_pol_file_rel = os.path.join("..", SCRIPTS_DIR, path_pol_path_tail)
-        return locals()
-
-    def zk_path_dict(self, isd_id, ad_id, zk_id):
+        self._delete_directories()
+        keys = self._write_keys_certs()
+        self._write_conf_files()
+        self._write_path_pol_files()
```
Comment (shitz) [inline topology/generator.py]: Can we name this `_write_path_policy_files()` please? Similar for `topo` below.

File: topology/ADConfigurations.json (hunk 1 above)
Comment (kormat) [inline topology/ADConfigurations.json]: Haha, nope. I'll change it.

File: topology/generator.py (hunk 2 above)
Comment (kormat) [inline topology/generator.py]: Yeah sure.

File: topology/generator.py [hunk 3]
```diff
@@ -359,7 +271,7 @@ def write_keys_certs(self, ad_configs):
                 'sig_pub_keys': sig_pub_keys,
                 'enc_pub_keys': enc_pub_keys}
 
-    def write_beginning_sim_run_files(self):
+    def _write_beginning_sim_run_files(self):
```
Comment (shitz) [inline topology/generator.py]: Please rename to `_write_sim_run_preamble()` or something like that.

File: topology/generator.py [hunk 4]
```diff
@@ -413,91 +320,88 @@ def write_topo_files(self, ad_configs, er_ip_addresses):
...
                     'AddrType': 'IPV4',
-                    'Addr': self.next_ip_address
+                    'Addr': str(addr),
                 }
-                self.next_ip_address = \
-                    self.increment_address(self.next_ip_address, mask)
             # Write Path Servers
-            if (ad_configs[isd_ad_id]['level'] != INTERMEDIATE_AD or
-                    "path_servers" in ad_configs[isd_ad_id]):
+            if (ad_conf['level'] != INTERMEDIATE_AD or
```
Comment (shitz) [inline topology/generator.py:352]: This condition seems a bit arbitrary to me. I don't think we should _disallow_ intermediate ASes to have path servers. I'd only do a `if "path_servers" in ad_conf:`

File: topology/generator.py (hunk 3 above)
Comment (kormat) [inline topology/generator.py]: Done, gladly.

File: topology/generator.py (hunk 4 above)
Comment (kormat) [inline topology/generator.py:352]: I would prefer if that was changed in a different PR (it also probably affects 2 other places in generator.py).

Comment (shitz) [inline topology/generator.py:352]: Ok. I see that the generator needs some major overhaul anyway. :+1:

File: topology/generator.py (hunk 3 above)
Comment (shitz) [inline topology/generator.py]: :+1:

File: topology/generator.py (hunk 2 above)
Comment (shitz) [inline topology/generator.py]: :+1:

File: topology/ADConfigurations.json (hunk 1 above)
Comment (shitz) [inline topology/ADConfigurations.json]: :+1:

File: topology/generator.py [hunk 5]
```diff
@@ -922,6 +782,26 @@ def dict_(self):
...
+        return self._map[location]
+
+
+class AddressGenerator(object):
+    def __init__(self, subnet):
+        self._subnet = subnet
+        self._hosts = self._subnet.hosts()
+        self._map = defaultdict(lambda: next(self._hosts))
+
+    def get(self, id_):
```
Comment (shitz) [inline topology/generator.py:811]: This should raise an exception when the generator cannot allocate more addresses within the subnet.

Comment (kormat) [inline topology/generator.py:811]: I've added error logging for these cases:

CRITICAL:root:Network 10.0.0.0/31 is too small to accomadate /27 subnets
CRITICAL:root:Unable to allocate any more subnets from 10.0.0.0/26
CRITICAL:root:Unable to allocate any more addresses from 10.0.0.0/31

Comment (shitz) [inline topology/generator.py:811]: :+1:

**Pull request merged**